### ISSUE TITLE
CESM+DART testing discussion

### ISSUE THREAD

**Issue by kdraeder** [ESMCI/cime#1807]
Before the CESM2 release, the DART project would like to have more
testing of DART-relevant parts of CESM included as standard procedure.
This would:
+ Enable evaluation of CESM model performance
  in the most recent CESM tag using DART
+ Minimize effort required to interface DART with future CESM tags.
   - Eliminate (minimize) SourceMods (except for specific user changes)
   - Eliminate modified CIME code from DART
   - Minimize debugging of CESM  and CIME code by  non-experts
+ Accelerate use of more recent CESM component versions
  in DART evaluations and science

Current DART interfaces to CESM include CAM (and related atms),
CLM, POP2, and CICE.  It appears to make most sense to start with
CAM+DART testing, then extend that to other components.
The DART team doesn't know much about CESM's testing procedures,
so we'll just describe what we think would be most helpful to
advancing data assimilation with CESM.  Then CSEG can decide what
can and should be done to accomplish that.

The fundamental feature DART needs is a robust multi-instance mode.
We think that at least one of the tests must use a 'full sized' 
ensemble: 80 instances of a 1 degree CAM6-FV, with no DART included.    
This test can make a very short forecast, like an hour, but will
require 240 nodes on Cheyenne.
Smaller ensembles (3-4 instances) of lower resolution could be used 
for some testing, but a full sized test is necessary.*
A smaller ensemble could be used to test the multi-cycle scripting,
in which at least 2 forecasts are made in the same job.
We think that it would also be helpful to test CAM variants which are
used with DART: WACCM(-X), CAM-SE, (MPAS? ...)

(* This will be especially true if/when the multi-coupler is brought
   into use.  Early versions were unusably slow for large ensembles
   because of redundant calls to build-namelist, which wasn't noticed
   in small ensemble tests.)

A second level of testing could be rolled into the multi-instance +  
multi-cycle testing, without involving DART executables and output, 
by using the no_assimilate.csh script.  That would test the CIME 
scripting involved with calling DART, including the multi-cycle 
job capability.

The most comprehensive testing would use DART executables to do
data assimilation for several cycles.  This would test the 
communication between CESM and DART, including the renaming of
DART files using the CESM convention, and short term archiving of them. 
It could also test DART's setup scripts, which use CIME scripts and
environment variables to create the case and stage the input files.
This is an important piece for DART users, and can be broken
by changes to CIME.
If full sized tests were already done, this could use a small ensemble.
The assimilations could be short; just a few observations of a variety
of types, so it would not take much time.

A variant of this test would be required if there are changes to CESM
components that change answers, and we want to know whether those
changes are significant, at least in the DART context.
Then we would want to use a full sized ensemble, standard resolution,
and possibly weeks of assimilation cycles.  But this may be more
of a science question, and not something to be done regularly
before tags.

Some preliminary work should be done before this testing is set up:
+ The current SourceMods needed by DART should be rolled into CESM.
+ Several modifications to CIME are also required.
  I'll open separate GIT issues to handle those.
+ Confirm with DART developers that the list of things to be tested
  is sufficient.

People potentially involved with this:
DART: 
-    Kevin Raeder
-    Nancy Collins
-    Tim Hoar

CESM
-    Mariana Vertenstein
-    Steve Goldhaber
-    Jim Edwards
-    Alice Bertini
-    ...?

Thanks for any thoughts about this!
Kevin, for DART

**Comment by gold2718** [comment]
I think this issue is a good place to work out the set of tests to add.
Any necessary new functionality should be split into one or more new issues.

**Comment by kdraeder** [comment]
@sacks mentioned in a CGD research report that he's working on CESM testing
requirements, and may be available to help set up the (python?) interface between
components and the testing software.  I'd like to start learning about that process
and figuring out the strategy for DART, while the merging of DART  SourceMods 
into the CESM trunk is going on.
Kevin

**Comment by jedwards4b** [comment]
Please tell me what advantage there is in an 80 instance test over an 8 instance one - or 2 instances for that matter?

**Comment by kdraeder** [comment]
Tests with just a few members would be useful,
but there have been several instances where a multi-instance
run worked fine for a few instances, but ran poorly or even
not at all for a large (O(80)) number.  I'd like to think that all
of those problems are in the past, but things get broken . . .
So at least one, short, large ensemble test is necessary
in order to convince us that a tag will be usable for answering
science questions in the DART context.  The large ensembe
test wouldn't necessarily need to be run, if there were no
changes to the multi-instance or multi-driver code (or ...?)
but we should have the ability to run it without a lot of extra
human effort.
